File: solutions/134/main.py
from typing import List 

class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        totalSum = 0
        curSum = 0
        result = 0
        for i in range(len(cost)):
            curSum += gas[i] - cost[i]
            totalSum += gas[i] - cost[i]
            if curSum < 0:
                curSum = 0
                result = i + 1
        if totalSum < 0:
            return -1
        return result

if __name__ == "__main__":
    gas = [1,2,3,4,5]
    cost = [3,4,5,1,2]
    mySol = Solution()
    assert mySol.canCompleteCircuit(gas, cost) == 3

File: solutions/134/main.cpp
#include <iostream>
#include <vector>
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    // time complexity O(n)
    // space complexity O(1)
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int total = 0, start = -1; 
        for (int i = 0, sum = 0; i < gas.size(); ++i)
        {
            sum += gas[i] - cost[i];
            total += gas[i] - cost[i]; 
            // if cannot go further at i, then set i as start, zero sum 
            if (sum < 0)
            {
                start = i; 
                sum = 0; 
            }
        }
        return total >= 0 ? start + 1 : -1; 
    }
};

TEST(Test134, SimpleTest)
{
    vector<int> gas = {1,2,3,4,5};
    vector<int> cost = {3,4,5,1,2}; 

    Solution s; 
    EXPECT_EQ(s.canCompleteCircuit(gas, cost), 3); 
}

int main()
{
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
